test/lifecycle_module_test.dart

```diff
@@ -948,20 +909,263 @@ void main() {
       ]);
     });
 
-    test('getManagedTimer should return a timer', () {
-      module.getManagedTimer(
-          new Duration(milliseconds: 10), expectAsync0(() {}));
-    });
+    group('disposal', () {
+      test('should be triggered by unload', () async {
+        await gotoState(module, LifecycleState.loaded);
+        await module.unload();
+        expect(module.isDisposed, isTrue);
+      });
+
+      test('should be a no-op if already disposing', () async {
+        var future = module.dispose();
+        await Future.wait([future, module.dispose()]);
+        expect(module.isDisposed, isTrue);
+      });
+
+      test('should be a no-op if already disposing via super.dispose()',
+          () async {
+        await gotoState(module, LifecycleState.loaded);
+        // ignore: unawaited_futures
+        module.dispose();
+        await module.didUnload.first;
+        await new Future(() {});
+        expect(module.isDisposing, isTrue);
+        await module.dispose();
+        expect(module.isDisposed, isTrue);
+      });
+
+      test('should render all API methods unusable', () async {
+        var invalidAfterDisposalMatcher = allOf(
+          throwsStateError,
+          throwsA(predicate((e) => e.toString().contains('dispos'))),
+        );
+
+        await module.dispose();
+        expect(module.load(), invalidAfterDisposalMatcher);
+        expect(module.loadChildModule(null), invalidAfterDisposalMatcher);
+        expect(module.resume(), invalidAfterDisposalMatcher);
+        expect(module.suspend(), invalidAfterDisposalMatcher);
+        expect(module.unload(), invalidAfterDisposalMatcher);
+      });
+
+      group('from instantiated state', () {
+        setUp(() async {
+          await gotoState(module, LifecycleState.instantiated);
+          module.eventList.clear();
+        });
+
+        test('should go straight to disposal', () async {
+          await module.dispose();
+          expect(module.isDisposed, isTrue);
+          expect(module.eventList, equals(['onDispose', 'didDispose']));
+        });
+
+        test('with an onDispose that throws', () async {
+          module.onDisposeError = testError;
+          expect(module.dispose(), throwsA(same(module.onDisposeError)));
+        });
+      });
 
-    test('getManagedPeriodicTimer should return a timer', () {
-      var callCount = 0;
-      module.getManagedPeriodicTimer(
-          new Duration(milliseconds: 10),
-          expectAsync1((Timer timer) {
-            if (callCount++ >= 1) {
-              timer.cancel();
-            }
-          }, count: 2));
+      void testDisposalFromLoadedState(LifecycleState state,
```

Comment:
should this return a future that's `awaitable` since it's performing async operations in its body?

Reply:
This method itself is not actually doing anything async (hence the `void` return type`) – it's simply defining a suite of tests via the `group()` and `test()` APIs based on the given parameters.